Implement letter frequency counting and printing

lexual opened the pull request:
Fixes #270

Review of colleague's PR #235 

#### Changes proposed in this PR:

## -

lexual-anz committed:
Implement letter frequency counting and printing

lexual-anz committed:
Use array instead of slices, where size is known

n0npax commented on 03_letters/lexual/freq.go:
```diff
@@ -0,0 +1,41 @@
+func sortLetters(m map[rune]int) []string {
+	keys := make([]rune, len(m))
+	i := 0
+	for k := range m {
```
`for i, k : range ... ` may be a better fit here

lexual replied:
isn't the behaviour that when you use range over a map the resulting pair is key,value not index,value ?

n0npax replied:
aww right, It's a map. sorry for that

n0npax commented on 03_letters/lexual/freq.go:
```diff
@@ -0,0 +1,41 @@
+		return keys[i] < keys[j]
+	})
+	sorted := make([]string, len(m))
+	for i, k := range keys {
```
I believe sorting should be done by `value`(occurrences) not by the `key`.

lexual replied:
The readme example seems to be sorted by key:

https://github.com/anz-bank/go-course/blob/master/03_letters/README.md

n0npax replied:
I just checked merged examples and they are done in the both ways, so it's fine.

camscale replied:
The readme example is sorted both by key (ascending) and by count (descending). Which should you do? Either way, you should document your assumptions.

rokane commented on 03_letters/lexual/freq.go:
```diff
@@ -0,0 +1,41 @@
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i] < keys[j]
+	})
+	sorted := make([]string, len(m))
```
By converting the map into a `[]rune` as the first step, we are introducing an intermediate step. Is there a need to convert to the `[]rune` first??? Perhaps we can create a `[]string` directly from the map, which will simplify this quite a lot. If that were the case, you could even perhaps use the `sort.Strings(input)` method to sort the slice, rather than providing your comparison function to the `sort.Slice()` method. 

Have a go at eliminating this `[]rune` and see if you can go directly to a `[]string`

lexual replied:
gone in new commit

rokane commented on 03_letters/lexual/freq.go:
```diff
@@ -0,0 +1,41 @@
+		keys[i] = k
+		i++
+	}
+	// sort keys
```
comment not required, please remove

lexual replied:
gone in new commit

rokane commented on 03_letters/lexual/freq_test.go:
```diff
@@ -0,0 +1,94 @@
+			},
+			[]string{"a:2", "b:1", "c:3"},
+		},
+		// empty map as argument tested
```
comment not required.

rokane commented on 03_letters/lexual/freq_test.go:
```diff
@@ -0,0 +1,94 @@
+			if v != actual[k] {
+				t.Errorf("Unexpected result")
+				t.Errorf("\nActual: %v\nExpected: %v", actual, test.expected)
+				for k, v := range actual {
```
I really don't think the looping over the `actual` and `expected` is beneficial here. It looks like you are doing this just for the purpose of printing, but you have already printed both the actual and expected above. This just introduces two iterations which are not required. Please remove this.

I would recommend running these test and providing an input which will make them fail. That way you will see what is printed and you will be able to determine if this is overkill. You will see only one `t.Errorf` should be required

lexual replied:
its because printing the map, shows the keys as integers instead of letters.

i.e. 

```Actual: map[97:1 98:2 99:3]
Expected: map[97:1 98:22 99:3]```

Isn't very useful for debugging, because I need to determine in my head that 97 corresponds to 'a'.

Full output gives this:

```freq_test.go:35: 
    Actual: map[97:1 98:2 99:3]
    Expected: map[97:1 98:22 99:3]
freq_test.go:37: Actual a = 1
freq_test.go:37: Actual b = 2
freq_test.go:37: Actual c = 3
freq_test.go:40: Expect c = 3
freq_test.go:40: Expect a = 1
freq_test.go:40: Expect b = 22
```

rokane commented on 03_letters/lexual/freq_test.go:
```diff
@@ -0,0 +1,94 @@
+			if actual[i] != test.expected[i] {
+				t.Errorf("actual and expected don't match")
+				t.Errorf("\nActual: %s\nExpected: %s", actual, test.expected)
+
```
Little nit here. But there is an extra line. Please remove this.

Also is there a reason why you are breaking here?

lexual replied:
gone in new commit

rokane reviewed (dismissed):
Looking good so far. I have left a hint for how to achieve the `sortLetters` function without the need for an intermediate slice. See if you can get this working, it will simplify and reduce your code a lot.

Can you also please update the description of the pull request so you are outlining the changes proposed in this piece of work. Thanks

lexual-anz committed:
Use []string instead of []rune

In response to code review, used []string
instead of []rune.

Also cleaned up the tests a little, and removed
unnecessary comments.

lexual-anz committed:
remote unnecessary conversion to string

lexual commented:
@rokane Can you follow up on PR. I believe I had responded to your past comments ;)

camscale commented on 03_letters/lexual/freq.go:
```diff
@@ -0,0 +1,38 @@
+	sort.Strings(keys)
+	sorted := make([]string, len(m))
+	for i, k := range keys {
+		sorted[i] = fmt.Sprintf("%s:%d", k, m[rune(k[0])])
```
A question for you to answer yourself - why do you need the `rune` cast here? what does that mean for the value returned by the `k[0]` expression? It is related to the test case addition I suggested. Hope I'm not being too obtuse - just trying to get you to think about this some more.

lexual replied:
adopted DecodeRuneInString ;)

camscale commented on 03_letters/lexual/freq_test.go:
```diff
@@ -0,0 +1,91 @@
+				'c': 3,
+			},
+			[]string{"a:2", "b:1", "c:3"},
+		},
```
You should add a test case with unicode characters that are encoded as multi-byte values as utf8. You might be surprised by the result (and then you can fix your code - you may want to look at `utf8.DecodeRuneInString()`)

camscale commented on 03_letters/lexual/freq_test.go:
```diff
@@ -0,0 +1,91 @@
+	}
+	for _, test := range tests {
+		actual := letters(test.input)
+		if len(actual) != len(test.expected) {
```
I feel this whole test is too verbose. It should be simple to understand what is being tested. You could just use `reflect.DeepEqual`. I understand your comment about the output being integers instead of characters, but you need to balance that with the readability of the test. If you really still want prettier output, I suggest adding a function that creates a string representation of the map as you want and using that in your error message. That keeps the test simple and moves the map formatting logic elsewhere.

lexual replied:
simplified

camscale reviewed (changes requested):
Your implementation is subtly wrong and has gone unnoticed because your test suite is too short. Have a read of https://blog.golang.org/strings for the gory details of strings which may help you know where you've gone wrong.

camscale commented:
Also, I noticed you asked for a re-review but you still had the "Changes Requested" tag and did not have the "Ready for Review" tag. We won't pick things up for review tagged like this.

lexual-anz committed:
lab03: use DecodeRuneInString & simply tests

Responding to @camh code review.

Have simplified the test code, and adopted use of
reflect.DeepEqual.

Now using utf8.DecodeRuneInString so multi-byte
runes are correctly handled.

camscale reviewed (changes requested):
The changes you said you made are not here. Did you forget to push?

lexual commented:
> The changes you said you made are not here. Did you forget to push?

D'oh! 

Yes, rebased and pushed just now ;(

camscale reviewed (approved):
LGTM.

When you merge this, you can either do a squash merge (give your commit a decent comment at that time) or you can rebase-squash-force-push down to a single commit and do a merge commit. I personally prefer to do the former (I like merge commits and I often have multiple commits in a PR, but I still need to squash fixups into the relevant commit). This is to avoid committing fixup commits to master to keep the history clean. If you do the latter, ping me on slack and I'll try to get to re-approving quickly (which is needed because a rebase-force-push needs re-approval) so you can merge.

juliaogris reviewed (dismissed):
Please [fix commit messages](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#commit-messages) with `git rebase -i HASH` and add a PR description.

anzdaddy commented:
Please merge or close.
